packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -69,6 +70,96 @@ class ToolbarItemsParentData extends ContainerBoxParentData<RenderBox> {
   String toString() => '${super.toString()}; shouldPaint=$shouldPaint';
 }
 
+
+/// A data class that allows the SelectionOverlay to delegate
+/// the loupe's positioning to the loupe itself, based on the
+/// info in [LoupeSelectionOverlayInfoBearer].
+@immutable
+class LoupeSelectionOverlayInfoBearer {
+  /// Construct a [LoupeSelectionOverlayInfoBearer] from raw values.
+  ///
+  /// It is recommended to use [LoupeSelectionOverlayInfoBearer._selectionControlDerrived]
+  /// in order to not manually re-calculate all the required parameters.
+  const LoupeSelectionOverlayInfoBearer({
+    required this.globalGesturePosition,
+    required this.handleRect,
+    required this.fieldBounds,
+    required this.currentLineBoundries,
+  });
+
+  factory LoupeSelectionOverlayInfoBearer._selectionControlDerrived({
+    required RenderEditable renderEditable,
+    required Offset globalGesturePosition,
+    required TextPosition currentTextPosition,
+  }) {
+    final Offset globalRenderEditableTopLeft = renderEditable.localToGlobal(Offset.zero);
+    final Rect localCaratRect = renderEditable.getLocalRectForCaret(currentTextPosition);
+
+
+    final TextSelection lineAtOffset = renderEditable.getLineAtOffset(currentTextPosition);
+    final TextPosition positionAtEndOfLine = TextPosition(
+        offset: lineAtOffset.extentOffset,
+        affinity: TextAffinity.upstream,
+    );
+
+    // Default affinity is downstream.
+    final TextPosition positionAtBeginningOfLine = TextPosition(
+      offset: lineAtOffset.baseOffset,
+    );
+
+    final Rect lineBoundries = Rect.fromPoints(
+      renderEditable.getLocalRectForCaret(positionAtBeginningOfLine).topLeft,
+      renderEditable.getLocalRectForCaret(positionAtEndOfLine).bottomRight
+    );
+
+
+    return LoupeSelectionOverlayInfoBearer(
+      fieldBounds: globalRenderEditableTopLeft & renderEditable.size,
+      globalGesturePosition: globalGesturePosition,
+      handleRect: localCaratRect.shift(globalRenderEditableTopLeft),
+      currentLineBoundries: lineBoundries.shift(globalRenderEditableTopLeft)
+    );
+  }
+
+  /// Construct an empty [LoupeSelectionOverlayInfoBearer], with all
+  /// values set to 0.
+  const LoupeSelectionOverlayInfoBearer.empty() :
+    globalGesturePosition = Offset.zero,
+    handleRect = Rect.zero,
+    currentLineBoundries = Rect.zero,
+    fieldBounds = Rect.zero;
+
+  /// The offset of the gesture position that the loupe should be shown at.
+  final Offset globalGesturePosition;
+
+  /// The rect of the current line the loupe should be shown at. Do not take
+  /// into account any padding of the field; only the position of the first
+  /// and last character.
+  final Rect currentLineBoundries;
```

Comment:
As a general rule, it's only a matter of time before someone needs more text layout information than whatever you think they'll need. For this reason, in super_editor, we typically pass around an entire `TextLayout` interface. I'm not sure that Flutter has one at this point, so you might consider handing over the `RenderParagraph` instead.

Reply:
Maybe that's also an argument in favor of LoupeBuilder's type parameter.  I was thinking that the type parameter was overdoing it and hardcoding it to LoupeSelectionOverlayInfoBearer would be fine, but actually, with the type parameter you could write your own "InfoBearer" class that does pass around RenderParagraph/RenderEditable or whatever.